File: extract/brickmesh_extract/connect.py
# SPDX-License-Identifier: Apache-2.0
# Copyright 2026 Sander Striker
"""
Connecting separate pieces.

Two parts can only join directly if their holes happen to coincide with the
same direction. In a real mechanism that is rarely so: the bearings carry
shafts that stand perpendicular to each other, so their holes point every which
way.

What you need is a CHAIN of intermediate parts carrying one hole direction over
to the other. That is a Steiner tree problem, and between two pieces it comes
down to a shortest path: every part added widens the set of reachable ports,
and you look for the cheapest route until you touch a port of the other piece.

A* with, as its heuristic, the remaining distance divided by how far a part
reaches at most. Without a heuristic the search wanders off: there are close to
a thousand placements per port.
"""
from __future__ import annotations

import heapq

import numpy as np

from .portsynth import ROUND, Placed, part_ports, rotations_mapping, world_ports
from .voxel import ROTATIONS, voxels

HALF_STUD = 10.0
MAX_REACH = 220.0          # longest part in the inventory, in LDU


def _port_key(p, a):
    return (tuple(np.round(p, 2)), tuple(np.round(np.abs(a), 3)))


def placements_at(point, axis, cat: dict, want_gender: str = "F",
                  want_kind=None, cap: int = 25) -> list[Placed]:
    """
    Placements where a port of the part lands on `point` with the axis along
    `axis`. want_gender 'F' looks for holes, 'M' looks for pins.
    """
    point = np.asarray(point, float)
    axis = np.asarray(axis, float); axis = axis / np.linalg.norm(axis)
    out, seen = [], set()
    for pid in cat:
        holes, pins = part_ports(pid, cat)
        table = holes if want_gender == "F" else pins
        if not len(table):
            continue
        n = 0
        for row in table:
            if want_kind is not None and row[6] != want_kind:
                continue
            for ri in rotations_mapping(row[3:6], axis):
                origin = point - ROTATIONS[ri] @ row[:3]
                if np.any(np.abs(origin / HALF_STUD
                                 - np.round(origin / HALF_STUD)) > 1e-6):
                    continue
                pl = Placed(pid, ri, tuple(np.round(origin, 3)))
                if pl in seen:
                    continue
                seen.add(pl); out.append(pl); n += 1
                if n >= cap:
                    break
            if n >= cap:
                break
    return out


def free_ports(parts: list[Placed], cat: dict) -> list[tuple]:
    """Every port of a piece, as (position, axis, kind, gender)."""
    out = []
    for p in parts:
        h, pn = world_ports(p, cat)
        for r in h:
            out.append((r[:3], r[3:6], int(r[6]), "F"))
        for r in pn:
            out.append((r[:3], r[3:6], int(r[6]), "M"))
    return out


def _cells(pl: Placed) -> frozenset:
    c = voxels(pl.part, pl.rot)
    shift = np.round(np.asarray(pl.origin, float) / 5.0).astype(int)
    return frozenset(map(tuple, (c + shift).tolist()))


CONTACT_FRACTION = 0.12


def _compatible(a: frozenset, b: frozenset) -> bool:
    shared = len(a & b)
    return shared == 0 or shared <= CONTACT_FRACTION * min(len(a), len(b))


def connect(comp_a: list[Placed], comp_b: list[Placed], cat: dict,
            blocked: frozenset = frozenset(), max_parts: int = 3,
            beam: int = 14, cost_fn=None) -> list[Placed] | None:
    """
    Cheapest chain of parts tying piece A to piece B.
    Returns the parts to add, or None.
    """
    cost_fn = cost_fn or (lambda pl: 1.0)

    ports_a = free_ports(comp_a, cat)
    ports_b = free_ports(comp_b, cat)
    if not ports_a or not ports_b:
        return None
    target_pts = np.array([p[0] for p in ports_b])
    target_axes = np.array([p[1] / np.linalg.norm(p[1]) for p in ports_b])
    target_keys = {_port_key(p[0], p[1]) for p in ports_b}

    base_cells = frozenset().union(*[_cells(p) for p in comp_a + comp_b]) | blocked

    HW = 6.0        # heuristic weight; without it A* is just BFS

    def h(pl: Placed) -> float:
        h_, pn = world_ports(pl, cat)
        pts = np.vstack([x[:3] for x in list(h_) + list(pn)]) if len(h_) or len(pn) else None
        if pts is None:
            return 1e9
        axs = np.vstack([x[3:6] for x in list(h_) + list(pn)])
        axs = axs / np.linalg.norm(axs, axis=1, keepdims=True)
        # Only ports sharing an axis can ever be connected. Measure the
        # distance without that distinction and the heuristic rewards exactly
        # those candidates sitting uselessly on top of the target with a hole
        # across it, and prunes away the perpendicular connectors that do solve
        # it.
        align = np.abs(axs @ target_axes.T)
        d = np.linalg.norm(pts[:, None, :] - target_pts[None, :, :], axis=2)
        d = np.where(align > 1 - 1e-6, d, np.inf)
        best = float(d.min())
        if not np.isfinite(best):
            # no axis matches at all: score on how close the directions come
            # to each other, since that is what a next link has to fix
            return HW * (1.0 + (1.0 - float(align.max())))
        return HW * best / MAX_REACH

    start_frontier = tuple(sorted(_port_key(p[0], p[1]) for p in ports_a))
    heap = [(0.0, 0, 0, tuple(), start_frontier)]
    seen = set()
    counter = 0

    while heap:
        _f, g, _, chain, frontier = heapq.heappop(heap)
        if frontier in seen:
            continue
        seen.add(frontier)

        if set(frontier) & target_keys:
            return list(chain)
        if len(chain) >= max_parts:
            continue

        used = frozenset().union(*[_cells(p) for p in chain]) if chain else frozenset()
        occupied = base_cells | used

        # Sort ports by distance to the target and expand only the nearest.
        # Truncating arbitrarily throws away precisely the useful ports.
        fr = sorted(frontier,
                    key=lambda k: float(np.min(np.linalg.norm(
                        target_pts - np.array(k[0]), axis=1))))
        cands = []
        for key in fr[:8]:
            pos, ax = np.array(key[0]), np.array(key[1])
            # pin into the reachable hole, or hole on hole with a pin between
            for pl in (placements_at(pos, ax, cat, "M", cap=14)
                       + placements_at(pos, ax, cat, "F", ROUND, cap=14)):
                if pl in chain:
                    continue
                if not _compatible(_cells(pl), occupied):
                    continue
                cands.append(pl)

        scored = sorted({c: h(c) for c in cands}.items(), key=lambda kv: kv[1])
        for pl, hv in scored[:beam]:
            ng = g + cost_fn(pl)
            hh, pn = world_ports(pl, cat)
            newkeys = {_port_key(r[:3], r[3:6]) for r in list(hh) + list(pn)}
            nf = tuple(sorted(set(frontier) | newkeys))
            counter += 1
            heapq.heappush(heap, (ng + hv, ng, counter, (*chain, pl), nf))

    return None


# --------------------------------------------------------------------------
# index: the difference between 7.7 ms and microseconds per query
# --------------------------------------------------------------------------

_INDEX: dict = {}


def _axis_key(a) -> tuple:
    """Holes have no direction: +Y and -Y are the same hole."""
    a = np.asarray(a, float)
    a = a / np.linalg.norm(a)
    for v in a:
        if abs(v) > 1e-9:
            if v < 0:
                a = -a
            break
    return tuple(np.round(a, 3))


def build_index(cat: dict) -> dict:
    """
    Turn the question around.

    The search keeps asking: which part has a hole at THIS point with THIS
    direction. Answering that by walking every part costs 7.7 ms. But the set
    of answers does not depend on the point - only on the direction. So compute
    once, for every combination of part, hole and rotation, which world
    direction comes out, and group on that.

    A query then becomes: look up the table, and per hit origin = point minus
    the pre-rotated hole position. That is one subtraction.
    """
    idx: dict = {}
    for pid, e in cat.items():
        for gender, arr in (("F", e["holes"]), ("M", e["pins"])):
            for rec in arr:
                pos = np.array(rec[:3], float)
                ax = np.array(rec[3:6], float)
                kind = int(rec[6])
                for ri, R in enumerate(ROTATIONS):
                    key = (_axis_key(R @ ax), gender, kind)
                    idx.setdefault(key, []).append((pid, ri, R @ pos))
    for k, v in idx.items():
        parts = np.array([x[0] for x in v], dtype=object)
        rots = np.array([x[1] for x in v], dtype=np.int16)
        offs = np.array([x[2] for x in v], dtype=np.float32)
        idx[k] = (parts, rots, offs)
    return idx


def index_for(cat: dict) -> dict:
    key = id(cat)
    if key not in _INDEX:
        _INDEX[key] = build_index(cat)
    return _INDEX[key]


def placements_indexed(point, axis, cat: dict, gender: str = "F",
                       kind: int | None = None) -> list[Placed]:
    """The same answer as placements_at, but through the index."""
    idx = index_for(cat)
    point = np.asarray(point, float)
    key_axis = _axis_key(axis)
    out = []
    kinds = (0, 1) if kind is None else (kind,)
    for k in kinds:
        hit = idx.get((key_axis, gender, k))
        if hit is None:
            continue
        parts, rots, offs = hit
        origins = point[None, :] - offs
        ok = np.all(np.abs(origins / 10.0 - np.round(origins / 10.0)) < 1e-4, axis=1)
        for pid, ri, o in zip(parts[ok], rots[ok], origins[ok]):
            out.append(Placed(pid, int(ri), tuple(np.round(o, 3))))
    return out


def placements_topk(point, axis, cat: dict, gender: str = "F",
                    kind: int | None = None, target_pts=None,
                    k: int = 24) -> list[Placed]:
    """
    The same lookup, but filter and score VECTORIZED and only build objects for
    the best k.

    That is where the time went. The index finds 2600 candidates in
    microseconds, and then it costs 16 milliseconds to turn them into 2600
    Python objects that are almost all thrown away again. Score first, build
    only what you keep.
    """
    idx = index_for(cat)
    point = np.asarray(point, float)
    key_axis = _axis_key(axis)
    kinds = (0, 1) if kind is None else (kind,)

    P, R_, O = [], [], []
    for kk in kinds:
        hit = idx.get((key_axis, gender, kk))
        if hit is None:
            continue
        parts, rots, offs = hit
        P.append(parts); R_.append(rots); O.append(offs)
    if not P:
        return []
    parts = np.concatenate(P); rots = np.concatenate(R_); offs = np.concatenate(O)

    origins = point[None, :] - offs
    ok = np.all(np.abs(origins / 10.0 - np.round(origins / 10.0)) < 1e-4, axis=1)
    if not ok.any():
        return []
    parts, rots, origins = parts[ok], rots[ok], origins[ok]

    if target_pts is not None and len(origins):
        tp = np.asarray(target_pts, float)
        d = np.linalg.norm(origins[:, None, :] - tp[None, :, :], axis=2).min(axis=1)
        order = np.argsort(d)[:k]
    else:
        order = np.arange(min(k, len(origins)))

    return [Placed(str(parts[i]), int(rots[i]), tuple(np.round(origins[i], 3)))
            for i in order]

File: internal/connect/connect.go
// SPDX-License-Identifier: Apache-2.0
// Copyright 2026 Sander Striker

// Package connect joins two loose pieces of a structure.
//
// Two parts can only meet directly when their ports coincide with the same
// direction. In a real mechanism that is rare: the bearings carry shafts
// perpendicular to each other, so their holes point every which way.
//
// What is needed is a CHAIN of intermediate parts carrying one port direction
// over to the other. That is a Steiner tree problem, and between two pieces it
// reduces to a shortest path: every part added widens the set of reachable
// ports, and the cheapest route is the one that first touches a port of the
// other piece.
//
// A* with, as its heuristic, the remaining distance divided by how far a part
// can reach. Without a heuristic the search wanders: there are close to a
// thousand placements per port.
package connect

import (
	"container/heap"
	"math"
	"sort"

	"brickmesh/internal/catalog"
	"brickmesh/internal/geom"
	"brickmesh/internal/voxel"
)

const (
	// MaxReach is the longest part in the inventory, in LDU.
	MaxReach = 220.0
	// heuristicWeight scales the estimate. Without it A* degenerates to
	// breadth-first and the frontier explodes.
	heuristicWeight = 6.0
	// contactFraction separates touching from intersecting, as in the
	// structural search.
	contactFraction = 0.12
	// frontierWidth is how many ports of a partial chain are expanded. Sorted
	// by distance to the target first, so this keeps the nearest rather than an
	// arbitrary handful.
	frontierWidth = 8
)

// Port is a connection point in world coordinates.
type Port struct {
	Pos    geom.Vec3
	Axis   geom.Vec3
	Kind   catalog.PortKind
	Gender catalog.Gender
}

// portKey identifies a port for the frontier. Holes have no direction, so the
// axis is made sign-free.
type portKey struct{ pos, axis geom.Vec3 }

func keyOf(pos, axis geom.Vec3) portKey {
	a := axis.Unit()
	for _, v := range [3]float64{a.X, a.Y, a.Z} {
		if math.Abs(v) > 1e-9 {
			if v < 0 {
				a = a.Scale(-1)
			}
			break
		}
	}
	return portKey{pos.Round(2), a.Round(3)}
}

// PortsOf returns a placed part's ports in world coordinates.
func PortsOf(p catalog.Placement) []Port {
	r := geom.Rotations[p.Rot]
	out := make([]Port, 0, len(p.Part.Ports))
	for _, port := range p.Part.Ports {
		out = append(out, Port{
			Pos:    r.Apply(port.Pos).Add(p.Origin),
			Axis:   r.Apply(port.Axis),
			Kind:   port.Kind,
			Gender: port.Gender,
		})
	}
	return out
}

// Options tunes the search.
type Options struct {
	MaxParts int
	Beam     int
	// Blocked is space the chain may not occupy, beyond the two pieces.
	Blocked map[geom.Cell]bool
	// Cost is what adding a part is worth. Defaults to one each.
	Cost func(catalog.Placement) float64
}

// Connect returns the cheapest chain of parts tying piece A to piece B, or nil
// when there is none within the limits.
func Connect(cat *catalog.Catalog, rast *voxel.Rasterizer,
	pieceA, pieceB []catalog.Placement, opts Options) ([]catalog.Placement, error) {

	if opts.MaxParts <= 0 {
		opts.MaxParts = 3
	}
	if opts.Beam <= 0 {
		opts.Beam = 14
	}
	if opts.Cost == nil {
		opts.Cost = func(catalog.Placement) float64 { return 1 }
	}

	portsA, portsB := allPorts(pieceA), allPorts(pieceB)
	if len(portsA) == 0 || len(portsB) == 0 {
		return nil, nil
	}

	s := &search{
		cat: cat, rast: rast, opts: opts,
		targets: portsB,
		goal:    map[portKey]bool{},
		seen:    map[string]bool{},
		cells:   map[catalog.Placement][]geom.Cell{},
	}
	for _, p := range portsB {
		s.goal[keyOf(p.Pos, p.Axis)] = true
	}
	base, err := s.occupiedBy(append(append([]catalog.Placement{}, pieceA...), pieceB...))
	if err != nil {
		return nil, err
	}
	for c := range opts.Blocked {
		base[c] = true
	}
	s.base = base

	start := make([]portKey, 0, len(portsA))
	for _, p := range portsA {
		start = append(start, keyOf(p.Pos, p.Axis))
	}
	return s.run(start)
}

func allPorts(piece []catalog.Placement) []Port {
	var out []Port
	for _, p := range piece {
		out = append(out, PortsOf(p)...)
	}
	return out
}

type search struct {
	cat     *catalog.Catalog
	rast    *voxel.Rasterizer
	opts    Options
	targets []Port
	goal    map[portKey]bool
	base    map[geom.Cell]bool
	seen    map[string]bool
	cells   map[catalog.Placement][]geom.Cell
	counter int
}

// node is one partial chain.
type node struct {
	f, g     float64
	order    int
	chain    []catalog.Placement
	frontier []portKey
}

type queue []*node

func (q queue) Len() int      { return len(q) }
func (q queue) Swap(i, j int) { q[i], q[j] = q[j], q[i] }
func (q queue) Less(i, j int) bool {
	if q[i].f != q[j].f {
		return q[i].f < q[j].f
	}
	return q[i].order < q[j].order
}
func (q *queue) Push(x any) { *q = append(*q, x.(*node)) }
func (q *queue) Pop() any {
	old := *q
	n := old[len(old)-1]
	*q = old[:len(old)-1]
	return n
}

func (s *search) run(start []portKey) ([]catalog.Placement, error) {
	q := &queue{{f: 0, g: 0, frontier: sortedKeys(start)}}
	heap.Init(q)

	for q.Len() > 0 {
		cur := heap.Pop(q).(*node)
		sig := signature(cur.frontier)
		if s.seen[sig] {
			continue
		}
		s.seen[sig] = true

		if s.reachesGoal(cur.frontier) {
			return cur.chain, nil
		}
		if len(cur.chain) >= s.opts.MaxParts {
			continue
		}
		if err := s.expand(q, cur); err != nil {
			return nil, err
		}
	}
	return nil, nil
}

func (s *search) reachesGoal(frontier []portKey) bool {
	for _, k := range frontier {
		if s.goal[k] {
			return true
		}
	}
	return false
}

// expand pushes every worthwhile extension of a chain.
func (s *search) expand(q *queue, cur *node) error {
	occupied, err := s.occupiedBy(cur.chain)
	if err != nil {
		return err
	}
	for c := range s.base {
		occupied[c] = true
	}

	cands, err := s.candidates(cur, occupied)
	if err != nil {
		return err
	}
	scored := make([]scoredCandidate, 0, len(cands))
	for _, c := range cands {
		scored = append(scored, scoredCandidate{placement: c, h: s.heuristic(c)})
	}
	sort.SliceStable(scored, func(i, j int) bool { return scored[i].h < scored[j].h })
	if len(scored) > s.opts.Beam {
		scored = scored[:s.opts.Beam]
	}

	for _, sc := range scored {
		g := cur.g + s.opts.Cost(sc.placement)
		s.counter++
		heap.Push(q, &node{
			f:        g + sc.h,
			g:        g,
			order:    s.counter,
			chain:    append(append([]catalog.Placement{}, cur.chain...), sc.placement),
			frontier: mergeFrontier(cur.frontier, sc.placement),
		})
	}
	return nil
}

type scoredCandidate struct {
	placement catalog.Placement
	h         float64
}

// candidates are the parts that can attach to the nearest ports of the chain:
// a pin into a reachable hole, or hole on hole with a pin between.
func (s *search) candidates(cur *node, occupied map[geom.Cell]bool) ([]catalog.Placement, error) {
	frontier := s.nearestFirst(cur.frontier)
	if len(frontier) > frontierWidth {
		frontier = frontier[:frontierWidth]
	}

	var out []catalog.Placement
	var buf []catalog.Placement
	for _, k := range frontier {
		for _, want := range []struct {
			gender catalog.Gender
			kind   catalog.PortKind
		}{
			{catalog.Male, catalog.Round},
			{catalog.Female, catalog.Round},
		} {
			buf = s.cat.Lookup(k.pos, k.axis, want.gender, want.kind, buf)
			for _, pl := range buf {
				if inChain(cur.chain, pl) {
					continue
				}
				ok, err := s.compatible(pl, occupied)
				if err != nil {
					return nil, err
				}
				if ok {
					out = append(out, pl)
				}
			}
		}
	}
	return out, nil
}

// nearestFirst sorts the frontier by distance to the closest target, so the
// ports most likely to lead somewhere are expanded. Truncating arbitrarily
// would throw away precisely the useful ones.
func (s *search) nearestFirst(frontier []portKey) []portKey {
	out := append([]portKey(nil), frontier...)
	dist := make(map[portKey]float64, len(out))
	for _, k := range out {
		best := math.Inf(1)
		for _, t := range s.targets {
			if d := k.pos.Sub(t.Pos).Len(); d < best {
				best = d
			}
		}
		dist[k] = best
	}
	sort.SliceStable(out, func(i, j int) bool { return dist[out[i]] < dist[out[j]] })
	return out
}

// heuristic estimates what is left to bridge.
//
// Only ports sharing an axis can ever be connected. Measuring distance without
// that distinction rewards exactly the candidates sitting uselessly on top of
// the target with a hole across it, and prunes away the perpendicular
// connectors that do solve it.
func (s *search) heuristic(pl catalog.Placement) float64 {
	ports := PortsOf(pl)
	if len(ports) == 0 {
		return math.Inf(1)
	}
	best := math.Inf(1)
	bestAlign := 0.0
	for _, p := range ports {
		pa := p.Axis.Unit()
		for _, t := range s.targets {
			align := math.Abs(pa.Dot(t.Axis.Unit()))
			if align > bestAlign {
				bestAlign = align
			}
			if align < 1-1e-6 {
				continue
			}
			if d := p.Pos.Sub(t.Pos).Len(); d < best {
				best = d
			}
		}
	}
	if math.IsInf(best, 1) {
		// No axis matches at all: score on how close the directions come, since
		// that is what a next link has to fix.
		return heuristicWeight * (1 + (1 - bestAlign))
	}
	return heuristicWeight * best / MaxReach
}

func (s *search) compatible(pl catalog.Placement, occupied map[geom.Cell]bool) (bool, error) {
	cells, err := s.cellsOf(pl)
	if err != nil {
		return false, err
	}
	shared := 0
	for _, c := range cells {
		if occupied[c] {
			shared++
		}
	}
	if shared == 0 {
		return true, nil
	}
	smaller := len(cells)
	if len(occupied) < smaller {
		smaller = len(occupied)
	}
	return float64(shared) <= contactFraction*float64(smaller), nil
}

func (s *search) occupiedBy(parts []catalog.Placement) (map[geom.Cell]bool, error) {
	out := map[geom.Cell]bool{}
	for _, p := range parts {
		cells, err := s.cellsOf(p)
		if err != nil {
			return nil, err
		}
		for _, c := range cells {
			out[c] = true
		}
	}
	return out, nil
}

func (s *search) cellsOf(p catalog.Placement) ([]geom.Cell, error) {
	if c, ok := s.cells[p]; ok {
		return c, nil
	}
	rel, err := s.rast.Voxels(p.Part.ID, int(p.Rot))
	if err != nil {
		return nil, err
	}
	shift := geom.Cell{
		X: int32(math.Round(p.Origin.X / voxel.Pitch)),
		Y: int32(math.Round(p.Origin.Y / voxel.Pitch)),
		Z: int32(math.Round(p.Origin.Z / voxel.Pitch)),
	}
	out := make([]geom.Cell, len(rel))
	for i, c := range rel {
		out[i] = c.Add(shift)
	}
	s.cells[p] = out
	return out, nil
}

func inChain(chain []catalog.Placement, p catalog.Placement) bool {
	for _, c := range chain {
		if c == p {
			return true
		}
	}
	return false
}

// mergeFrontier adds a part's ports to the reachable set.
func mergeFrontier(frontier []portKey, pl catalog.Placement) []portKey {
	seen := make(map[portKey]bool, len(frontier))
	out := append([]portKey(nil), frontier...)
	for _, k := range frontier {
		seen[k] = true
	}
	for _, p := range PortsOf(pl) {
		k := keyOf(p.Pos, p.Axis)
		if !seen[k] {
			seen[k] = true
			out = append(out, k)
		}
	}
	return sortedKeys(out)
}

func sortedKeys(keys []portKey) []portKey {
	out := append([]portKey(nil), keys...)
	sort.SliceStable(out, func(i, j int) bool { return lessKey(out[i], out[j]) })
	return out
}

func lessKey(a, b portKey) bool {
	for _, pair := range [][2]float64{
		{a.pos.X, b.pos.X}, {a.pos.Y, b.pos.Y}, {a.pos.Z, b.pos.Z},
		{a.axis.X, b.axis.X}, {a.axis.Y, b.axis.Y}, {a.axis.Z, b.axis.Z},
	} {
		if pair[0] != pair[1] {
			return pair[0] < pair[1]
		}
	}
	return false
}

// signature identifies a frontier so the same reachable set is not explored
// twice.
func signature(keys []portKey) string {
	b := make([]byte, 0, len(keys)*16)
	for _, k := range keys {
		b = appendFloat(b, k.pos.X)
		b = appendFloat(b, k.pos.Y)
		b = appendFloat(b, k.pos.Z)
		b = appendFloat(b, k.axis.X)
		b = appendFloat(b, k.axis.Y)
		b = appendFloat(b, k.axis.Z)
	}
	return string(b)
}

func appendFloat(b []byte, v float64) []byte {
	bits := math.Float64bits(v)
	for i := 0; i < 8; i++ {
		b = append(b, byte(bits>>(8*i)))
	}
	return b
}
